vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/dataview/CheckboxGroupListDataViewImpl.java

```diff
@@ -0,0 +1,152 @@
+package com.vaadin.flow.component.checkbox.dataview;
+
+import com.vaadin.flow.component.checkbox.CheckboxGroup;
+import com.vaadin.flow.data.provider.*;
+import com.vaadin.flow.function.SerializableComparator;
+import com.vaadin.flow.function.SerializableConsumer;
+import com.vaadin.flow.function.SerializableFunction;
+import com.vaadin.flow.function.SerializablePredicate;
+import com.vaadin.flow.shared.Registration;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class CheckboxGroupListDataViewImpl<T> implements CheckboxGroupListDataView<T> {
+
+    private int itemsSize = 0;
+    private Set<SizeChangeListener> sizeChangeListeners;
+    private CheckboxGroup<T> component;
+
+    public CheckboxGroupListDataViewImpl(CheckboxGroup<T> checkboxGroup) {
+        this.component = checkboxGroup;
+    }
+
+    @Override
+    public T getNextItem(T item) {
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex != -1 && itemIndex < itemsList.size() - 1) {
+                return itemsList.get(itemIndex + 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasNextItem(T item) {
+        return getNextItem(item) != null;
+    }
+
+    @Override
+    public T getPreviousItem(T item) {
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex > 0) {
+                return itemsList.get(itemIndex - 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasPreviousItem(T item) {
+        return getPreviousItem(item) != null;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withFilter(SerializablePredicate<T> filter) {
+        CheckboxGroupListDataView<T> dataView = withFilterOrOrder(
+                dataProvider -> dataProvider.setFilter(filter));
+        // fire size change event if size has changed
+        getDataSize();
+        return dataView;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withSortComparator(SerializableComparator<T> sortComparator) {
+        return withFilterOrOrder(dataProvider -> dataProvider.setSortComparator(sortComparator));
+    }
+
+    @Override
+    public Stream<T> getAllItems() {
+        DataProvider<T, ?> dataProvider = getDataProvider();
+        return dataProvider.fetch(new Query<>());
```

Comment:
Should this take filtering into account ? My initial assumption is that it should based on the use cases we had.
Another thing is that should we have separate API for the list/inmemory case that provides all items even without the list ? I don't know why we would need that though

Reply:
Yes, we could add two overloaded methods with a SerializablePredicate<T> parameter and Query<T, SerializablePredicate<T>> parameter.

Reply:
> Another thing is that should we have separate API for the list/inmemory case that provides all items even without the list ?

Could you please describe in more details WDYM?

Reply:
The extra methods have been created to separate filtered and full data set.
- `getAllItems()` returns a full data set with no filters applied.
- `getItems()` returns a filtered data set.
- `getItems(Query)` returns a filtered (custom + query filter) data
- `getItems(SerializablePredicate)` returns a filtered data (custom + predicate filter), this is only for list data views.